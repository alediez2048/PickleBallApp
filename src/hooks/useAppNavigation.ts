import { useRouter } from 'expo-router';

export function useAppNavigation() {
  const router = useRouter();

  return {
    navigate: (route: string) => {
      router.push(route as any);
    },
    goBack: () => {
      router.back();
    }
  };
} 