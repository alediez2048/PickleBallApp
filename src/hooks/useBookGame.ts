import { useState, useCallback } from 'react';
import { GamesApi } from '@/services/api/games';
import { prefetch } from '@/utils/prefetch';
import type { Game } from '@/types/game';

interface BookingState {
  step: 'details' | 'payment' | 'confirmation';
  game: Game | null;
  isLoading: boolean;
  error: Error | null;
}

export function useBookGame(gameId?: string) {
  const [state, setState] = useState<BookingState>({
    step: 'details',
    game: null,
    isLoading: false,
    error: null,
  });

  const gamesApi = new GamesApi();

  const loadGame = useCallback(async () => {
    if (!gameId) return;

    setState(prev => ({ ...prev, isLoading: true, error: null }));
    try {
      const game = await gamesApi.getGame(gameId);
      setState(prev => ({ ...prev, game, isLoading: false }));
    } catch (err) {
      const error = err instanceof Error ? err : new Error('Failed to load game');
      setState(prev => ({ ...prev, error, isLoading: false }));
    }
  }, [gameId]);

  const startBooking = useCallback(async (game: Game) => {
    setState(prev => ({ 
      ...prev,
      step: 'details',
      game,
      error: null,
    }));

    // Prefetch data needed for the next steps
    prefetch.prefetchBookingFlow(game.id);
  }, []);

  const goToPayment = useCallback(() => {
    setState(prev => ({ ...prev, step: 'payment' }));
  }, []);

  const goToConfirmation = useCallback(() => {
    setState(prev => ({ ...prev, step: 'confirmation' }));
  }, []);

  const resetBooking = useCallback(() => {
    setState({
      step: 'details',
      game: null,
      isLoading: false,
      error: null,
    });
  }, []);

  return {
    ...state,
    loadGame,
    startBooking,
    goToPayment,
    goToConfirmation,
    resetBooking,
  };
} 