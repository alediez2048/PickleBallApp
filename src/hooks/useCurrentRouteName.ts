import { usePathname } from "expo-router";

/**
 * Returns a stable route "name" derived from the current pathname.
 * Optimized for the tabs in `(tabs)`: index | explore | profile | admin.
 * Falls back to the last segment for non-tab routes.
 */
export function useCurrentRouteName(): string | undefined {
  const pathname = usePathname();
  if (!pathname) return undefined;

  // Normalize by stripping query/hash and trailing slashes
  const cleaned = pathname.split("?")[0].split("#")[0].replace(/\/$/, "");

  if (cleaned === "" || cleaned === "/") return "index";
  if (cleaned.startsWith("/explore")) return "explore";
  if (cleaned.startsWith("/profile")) return "profile";
  if (cleaned.startsWith("/admin")) return "admin";

  // Fallback: return last segment as a route-ish name
  const parts = cleaned.replace(/^\/+/, "").split("/");
  return parts[parts.length - 1] || "index";
}
