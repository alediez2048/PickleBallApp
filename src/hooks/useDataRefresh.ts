import { useEffect, useRef, useState } from 'react';
import { useNetInfo } from '@react-native-community/netinfo';

interface RefreshConfig {
  retryAttempts?: number;
  retryDelay?: number;
  onSuccess?: () => void;
  onError?: (error: Error) => void;
}

const DEFAULT_CONFIG: Required<RefreshConfig> = {
  retryAttempts: 3,
  retryDelay: 1000,
  onSuccess: () => {},
  onError: () => {},
};

export function useDataRefresh<T>(
  key: string,
  fetcher: () => Promise<T>,
  config: RefreshConfig = {}
) {
  const { isConnected } = useNetInfo();
  const [isRefreshing, setIsRefreshing] = useState(false);
  const [error, setError] = useState<Error | null>(null);
  const retryCount = useRef(0);
  const timeoutRef = useRef<NodeJS.Timeout>();

  const fullConfig = { ...DEFAULT_CONFIG, ...config };

  const refresh = async () => {
    if (!isConnected) {
      setError(new Error('No internet connection'));
      return;
    }

    setIsRefreshing(true);
    setError(null);

    try {
      await fetcher();
      retryCount.current = 0;
      fullConfig.onSuccess();
    } catch (err) {
      const error = err instanceof Error ? err : new Error('Unknown error occurred');
      setError(error);

      if (retryCount.current < fullConfig.retryAttempts) {
        timeoutRef.current = setTimeout(() => {
          retryCount.current += 1;
          refresh();
        }, fullConfig.retryDelay * Math.pow(2, retryCount.current)); // Exponential backoff
      } else {
        fullConfig.onError(error);
      }
    } finally {
      setIsRefreshing(false);
    }
  };

  useEffect(() => {
    const handleRefresh = (event: Event) => {
      const customEvent = event as CustomEvent<{ key: string }>;
      if (customEvent.detail.key === key) {
        refresh();
      }
    };

    window.addEventListener('cacheRefresh', handleRefresh);

    return () => {
      window.removeEventListener('cacheRefresh', handleRefresh);
      if (timeoutRef.current) {
        clearTimeout(timeoutRef.current);
      }
    };
  }, [key]);

  return {
    isRefreshing,
    error,
    refresh,
  };
} 