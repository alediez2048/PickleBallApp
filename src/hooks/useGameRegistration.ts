import { useState, useEffect, useCallback } from 'react';
import { mockApi } from '@/services/mockApi';
import { GAME_CONSTANTS } from '@/types/game';
import { MOCK_GAMES } from '@/utils/mockData';

interface GameRegistrationState {
  registeredCount: number;
  isLoading: boolean;
  error: Error | null;
  isFull: boolean;
  spotsLeft: number;
}

export function useGameRegistration(gameId: string) {
  const [state, setState] = useState<GameRegistrationState>({
    registeredCount: MOCK_GAMES[gameId]?.registeredCount || 0,
    isLoading: true,
    error: null,
    isFull: false,
    spotsLeft: GAME_CONSTANTS.MAX_PLAYERS,
  });

  const fetchRegistrationCount = useCallback(async () => {
    if (!gameId) {
      console.error('No gameId provided to useGameRegistration');
      setState(prev => ({
        ...prev,
        isLoading: false,
        error: new Error('Invalid game ID'),
      }));
      return;
    }

    try {
      // First try to get the count from mock data as fallback
      const mockGame = MOCK_GAMES[gameId];
      const initialCount = mockGame?.registeredCount || 0;

      // Then get the real-time count
      const count = await mockApi.getGameBookings(gameId);
      const finalCount = count || initialCount; // Use mock data if API returns 0
      const spotsLeft = GAME_CONSTANTS.MAX_PLAYERS - finalCount;

      setState({
        registeredCount: finalCount,
        isLoading: false,
        error: null,
        isFull: finalCount >= GAME_CONSTANTS.MAX_PLAYERS,
        spotsLeft,
      });
    } catch (error) {
      console.error('Error fetching registration count:', error);
      
      // Fallback to mock data if available
      const mockGame = MOCK_GAMES[gameId];
      if (mockGame) {
        const count = mockGame.registeredCount;
        const spotsLeft = GAME_CONSTANTS.MAX_PLAYERS - count;
        setState({
          registeredCount: count,
          isLoading: false,
          error: null,
          isFull: count >= GAME_CONSTANTS.MAX_PLAYERS,
          spotsLeft,
        });
      } else {
        setState(prev => ({
          ...prev,
          isLoading: false,
          error: error instanceof Error ? error : new Error('Failed to fetch registration count'),
        }));
      }
    }
  }, [gameId]);

  // Set up polling for real-time updates
  useEffect(() => {
    fetchRegistrationCount();

    const intervalId = setInterval(() => {
      fetchRegistrationCount();
    }, GAME_CONSTANTS.DEFAULT_POLLING_INTERVAL);

    return () => clearInterval(intervalId);
  }, [fetchRegistrationCount]);

  const formatSpotsMessage = useCallback(() => {
    if (state.isLoading) return 'Loading...';
    if (state.error) {
      // If we have a fallback count, show it even if there's an error
      if (state.registeredCount > 0) {
        return `${GAME_CONSTANTS.MAX_PLAYERS - state.registeredCount}/${GAME_CONSTANTS.MAX_PLAYERS} spots left`;
      }
      return 'Unable to load spots';
    }
    if (state.isFull) return 'Game Full';
    return `${state.spotsLeft}/${GAME_CONSTANTS.MAX_PLAYERS} spots left`;
  }, [state.isLoading, state.error, state.isFull, state.spotsLeft, state.registeredCount]);

  return {
    ...state,
    formatSpotsMessage,
    refresh: fetchRegistrationCount,
  };
} 