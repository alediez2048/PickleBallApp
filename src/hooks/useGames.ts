import { useQuery } from '@tanstack/react-query';
import { Game } from '@/types/game';
import { GamesApi, GameFiltersType } from '@/services/api/games';
import { useState, useEffect } from 'react';
import { mockApi } from '@/services/mockApi';
import { useUserProfile } from '@/contexts/selectors/authSelectors';

export function useGames(filters?: GameFiltersType) {
  const {
    data: games,
    isLoading,
    error,
    refetch,
  } = useQuery({
    queryKey: ['games', filters],
    queryFn: () => GamesApi.getGames(filters),
    staleTime: 1000 * 60 * 5, // 5 minutes
  });

  const refreshGames = async () => {
    await refetch();
  };

  return {
    games: games || [],
    isLoading,
    error,
    isRefreshing: isLoading,
    refreshGames,
  };
}

interface GameHistory {
  id: string;
  date: string;
  result: 'win' | 'loss';
  score: string;
  opponent: string;
}

export function useGameHistory() {
  const user = useUserProfile();
  const [games, setGames] = useState<GameHistory[]>([]);
  const [isLoading, setIsLoading] = useState(true);
  const [error, setError] = useState<string | null>(null);

  useEffect(() => {
    async function fetchGames() {
      if (!user?.email) return;
      
      try {
        setIsLoading(true);
        const gameHistory = await mockApi.getGameHistory(user.email);
        setGames(gameHistory);
        setError(null);
      } catch (err) {
        setError('Failed to load game history');
        console.error('Error fetching game history:', err);
      } finally {
        setIsLoading(false);
      }
    }

    fetchGames();
  }, [user?.email]);

  const addGame = async (game: Omit<GameHistory, 'id'>) => {
    if (!user?.email) return;

    try {
      setIsLoading(true);
      const newGame = await mockApi.addGameToHistory(user.email, game);
      setGames(prevGames => [newGame, ...prevGames]);
      setError(null);
    } catch (err) {
      setError('Failed to add game');
      console.error('Error adding game:', err);
    } finally {
      setIsLoading(false);
    }
  };

  return {
    games,
    isLoading,
    error,
    addGame
  };
} 