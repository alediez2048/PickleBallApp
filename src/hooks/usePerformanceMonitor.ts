import { useRef, useEffect } from 'react';
import { InteractionManager } from 'react-native';

export interface PerformanceMetrics {
  mountTime: number;
  renderTime: number;
  interactionTime: number;
}

interface PerformanceMonitorOptions {
  componentName: string;
  onMetricsCollected?: (metrics: PerformanceMetrics) => void;
}

export function usePerformanceMonitor({ componentName, onMetricsCollected }: PerformanceMonitorOptions) {
  const startTimeRef = useRef<number>(0);
  const mountTimeRef = useRef<number>(0);
  const renderTimeRef = useRef<number>(0);

  useEffect(() => {
    startTimeRef.current = performance.now();

    const interaction = InteractionManager.createInteractionHandle();

    return () => {
      const endTime = performance.now();
      const metrics: PerformanceMetrics = {
        mountTime: mountTimeRef.current - startTimeRef.current,
        renderTime: renderTimeRef.current - startTimeRef.current,
        interactionTime: endTime - startTimeRef.current,
      };

      console.debug(`[Performance] ${componentName}:`, metrics);
      onMetricsCollected?.(metrics);

      InteractionManager.clearInteractionHandle(interaction);
    };
  }, [componentName, onMetricsCollected]);

  useEffect(() => {
    mountTimeRef.current = performance.now();
  }, []);

  useEffect(() => {
    renderTimeRef.current = performance.now();
  });

  return null;
} 