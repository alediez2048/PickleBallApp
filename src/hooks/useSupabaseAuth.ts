import { useState, useEffect, useCallback } from 'react';
import { Session, User, RealtimeChannel } from '@supabase/supabase-js';
import * as authService from '@/services/api/supabase/auth';
import { DBProfile } from '@/services/api/supabase/schema';

type AuthSubscription = {
  data: { subscription: { unsubscribe: () => void } };
};

/**
 * Custom hook for Supabase authentication
 */
export function useSupabaseAuth() {
  const [user, setUser] = useState<User | null>(null);
  const [session, setSession] = useState<Session | null>(null);
  const [profile, setProfile] = useState<DBProfile | null>(null);
  const [isLoading, setIsLoading] = useState<boolean>(true);
  const [error, setError] = useState<Error | null>(null);

  // Initialize auth and session
  useEffect(() => {
    let mounted = true;
    let authListener: AuthSubscription | null = null;
    let profileListener: RealtimeChannel | null = null;

    async function initialize() {
      try {
        // Initialize auth service
        await authService.initializeAuth();

        // Get initial session
        const { session: initialSession, error: sessionError } = await authService.getSession();
        if (sessionError) throw sessionError;

        if (mounted) {
          setSession(initialSession);
          
          if (initialSession?.user) {
            setUser(initialSession.user);
            
            // Fetch profile
            profileListener = await fetchProfile(initialSession.user.id);
          }
        }

        // Set up auth state change listener
        authListener = authService.onAuthStateChange((event, changedSession) => {
          if (mounted) {
            setSession(changedSession);
            setUser(changedSession?.user || null);
            
            if (changedSession?.user) {
              // Clean up previous listener if it exists
              if (profileListener) {
                profileListener.unsubscribe();
              }
              
              // Set up new listener
              fetchProfile(changedSession.user.id).then(listener => {
                if (listener) {
                  profileListener = listener;
                }
              });
            } else if (event === 'SIGNED_OUT') {
              setProfile(null);
            }
          }
        }) as AuthSubscription;
      } catch (err) {
        if (mounted) {
          setError(err instanceof Error ? err : new Error(String(err)));
        }
      } finally {
        if (mounted) {
          setIsLoading(false);
        }
      }
    }

    initialize();

    return () => {
      mounted = false;
      // Clean up listeners
      if (authListener?.data?.subscription) {
        authListener.data.subscription.unsubscribe();
      }
      if (profileListener) {
        profileListener.unsubscribe();
      }
    };
  }, []);

  // Fetch user profile
  const fetchProfile = useCallback(async (userId: string): Promise<RealtimeChannel | null> => {
    try {
      const { profile: userProfile, error: profileError } = await authService.getUserProfile(userId);
      if (profileError) throw profileError;

      setProfile(userProfile);

      // Set up profile change listener
      const listener = authService.onProfileChange(userId, (updatedProfile) => {
        if (updatedProfile) {
          setProfile(updatedProfile);
        }
      });
      
      return listener;
    } catch (err) {
      console.error('Error fetching profile:', err);
      setError(err instanceof Error ? err : new Error(String(err)));
      return null;
    }
  }, []);

  // Sign in with email and password
  const signIn = useCallback(async (email: string, password: string) => {
    setIsLoading(true);
    setError(null);

    try {
      const { user: authUser, session: authSession, error: authError } = 
        await authService.signIn({ email, password });

      if (authError) throw authError;

      setUser(authUser);
      setSession(authSession);

      if (authUser) {
        await fetchProfile(authUser.id);
      }

      return { user: authUser, error: null };
    } catch (err) {
      const errorMessage = err instanceof Error ? err.message : String(err);
      setError(new Error(errorMessage));
      return { user: null, error: new Error(errorMessage) };
    } finally {
      setIsLoading(false);
    }
  }, [fetchProfile]);

  // Sign up with email and password
  const signUp = useCallback(async (email: string, password: string, name: string) => {
    setIsLoading(true);
    setError(null);

    try {
      const { user: authUser, error: authError } = 
        await authService.signUp({ email, password, name });

      if (authError) throw authError;

      return { user: authUser, error: null };
    } catch (err) {
      const errorMessage = err instanceof Error ? err.message : String(err);
      setError(new Error(errorMessage));
      return { user: null, error: new Error(errorMessage) };
    } finally {
      setIsLoading(false);
    }
  }, []);

  // Sign in with a third-party provider
  const signInWithProvider = useCallback(async (provider: 'google' | 'facebook' | 'apple') => {
    setIsLoading(true);
    setError(null);

    try {
      const { error: authError } = 
        await authService.signInWithProvider({ provider });

      if (authError) throw authError;

      return { error: null };
    } catch (err) {
      const errorMessage = err instanceof Error ? err.message : String(err);
      setError(new Error(errorMessage));
      return { error: new Error(errorMessage) };
    } finally {
      setIsLoading(false);
    }
  }, []);

  // Sign out
  const signOut = useCallback(async () => {
    setIsLoading(true);
    setError(null);

    try {
      const { error: authError } = await authService.signOut();

      if (authError) throw authError;

      setUser(null);
      setSession(null);
      setProfile(null);

      return { error: null };
    } catch (err) {
      const errorMessage = err instanceof Error ? err.message : String(err);
      setError(new Error(errorMessage));
      return { error: new Error(errorMessage) };
    } finally {
      setIsLoading(false);
    }
  }, []);

  // Reset password
  const resetPassword = useCallback(async (email: string) => {
    setIsLoading(true);
    setError(null);

    try {
      const { error: authError } = await authService.resetPassword({ email });

      if (authError) throw authError;

      return { error: null };
    } catch (err) {
      const errorMessage = err instanceof Error ? err.message : String(err);
      setError(new Error(errorMessage));
      return { error: new Error(errorMessage) };
    } finally {
      setIsLoading(false);
    }
  }, []);

  // Update password
  const updatePassword = useCallback(async (password: string) => {
    setIsLoading(true);
    setError(null);

    try {
      const { error: authError } = await authService.updatePassword({ password });

      if (authError) throw authError;

      return { error: null };
    } catch (err) {
      const errorMessage = err instanceof Error ? err.message : String(err);
      setError(new Error(errorMessage));
      return { error: new Error(errorMessage) };
    } finally {
      setIsLoading(false);
    }
  }, []);

  return {
    user,
    session,
    profile,
    isLoading,
    error,
    signIn,
    signUp,
    signInWithProvider,
    signOut,
    resetPassword,
    updatePassword,
    refreshProfile: () => user && fetchProfile(user.id)
  };
} 