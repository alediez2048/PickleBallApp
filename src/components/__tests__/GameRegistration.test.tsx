import React from 'react';
import renderer from 'react-test-renderer';
import { Text, View, TouchableOpacity } from 'react-native';

// Mock GAME_CONSTANTS to avoid reference error
jest.mock('@/types/game', () => ({
  GAME_CONSTANTS: {
    MAX_PLAYERS: 12,
    DEFAULT_POLLING_INTERVAL: 30000
  }
}));

// Mock for default state
jest.mock('@/hooks/useGameRegistration', () => {
  return {
    // Default state with available spots
    useGameRegistration: jest.fn().mockImplementation(() => ({
      registeredCount: 5,
      isLoading: false,
      error: null,
      isFull: false,
      spotsLeft: 7,
      formatSpotsMessage: () => '7/12 spots left',
      refresh: jest.fn(),
    })),
  };
});

// Import after mocking
import { useGameRegistration } from '@/hooks/useGameRegistration';

// Mock the expo-router
jest.mock('expo-router', () => ({
  router: {
    push: jest.fn(),
    replace: jest.fn(),
    back: jest.fn(),
  },
  useRouter: () => ({
    push: jest.fn(),
    replace: jest.fn(),
    back: jest.fn(),
  }),
  useLocalSearchParams: () => ({ id: 'game-123' }),
}));

// Mock the useAuth hook
jest.mock('@/contexts/AuthContext', () => ({
  useAuth: jest.fn(() => ({
    user: { id: 'user-123', hasCompletedProfile: true },
    signIn: jest.fn(),
    signUp: jest.fn(),
    signOut: jest.fn(),
  })),
}));

// Mock the API
jest.mock('@/services/mockApi', () => ({
  mockApi: {
    getGame: jest.fn(() => ({
      id: 'game-123',
      title: 'Test Game',
      date: new Date(),
      location: 'Test Location',
      skillLevel: 'Intermediate',
      price: 15,
      maxPlayers: 12,
    })),
    bookGame: jest.fn().mockResolvedValue({ success: true }),
    getGameBookings: jest.fn(() => 5),
    cancelBooking: jest.fn().mockResolvedValue({ success: true }),
  },
}));

// Define props interface for the component
interface GameRegistrationProps {
  gameId?: string;
}

// Create a simple mock for the components we're testing
const GameRegistration: React.FC<GameRegistrationProps> = ({ gameId = 'game-123' }) => {
  const gameRegistration = useGameRegistration(gameId);
  const { formatSpotsMessage, isLoading, isFull } = gameRegistration;

  return (
    <View testID="game-registration">
      <Text testID="spots-message">{formatSpotsMessage()}</Text>
      {isLoading && <Text testID="loading-indicator">Loading...</Text>}
      {isFull && <Text testID="full-message">Game is full</Text>}
      <TouchableOpacity 
        testID="register-button"
        disabled={isFull}
        onPress={() => {/* would handle registration */}}
      >
        <Text>Register</Text>
      </TouchableOpacity>
    </View>
  );
};

describe('GameRegistration', () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('renders correctly with default spots availability', () => {
    const tree = renderer
      .create(<GameRegistration gameId="game-123" />)
      .toJSON();
    
    expect(tree).toMatchSnapshot();
  });
}); 