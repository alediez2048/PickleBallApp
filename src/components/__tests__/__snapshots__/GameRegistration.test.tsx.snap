// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GameRegistration renders correctly with default spots availability 1`] = `null`;
